contents/javascript/Javascript-promises.md

```diff
@@ -0,0 +1,164 @@
+
+We can instead use Promises for cleaner code that would be easier to write and debug.
+
+## What is a Promise?
+
+The following example is adapted from [JavaScript Promises for Dummies](https://scotch.io/tutorials/javascript-promises-for-dummies).
+
+Promises in Javascript behave the same way as Promises do in real life. Imagine that your friend promises to return you money that you have lent him. At the time this promise was made, you would not know if your friend would really return your money. Your friend could either return your money on time, or he could just not do it.
+These situations correspond to the 3 states of Promises in Javascript.
+
+| State | Description |
+| ------ | ----------- |
+|Pending | You don't know if he would return your money. |
+|Fulfilled | He returned your money. |
+|Rejected | He refused to return your money. |
+
+
+Promises provide the power to control the execution flow of your program by allowing you to wait for a asynchronous operation to complete before carrying on with the rest of the program execution.
+
+## How Promises are used
+
+There are many interesting ways to use Promises. The following examples are written with Node.js.
+
+### HTTP Requests
+
+Earlier in this chapter, I mentioned how HTTP requests are asynchronous operations. By using Promises, you would be able to act on the result from the request without having to use callbacks or wait an arbitrarily set amount of time for the response to be returned.
+
+The following code sends a request to an Amazon product URL and retrieves the price of the product using Promises. By using Promises instead of callbacks, we have a nice little web scraping snippet that's easy for anyone to use and to debug if any issues arise.
+
+```javascript
+
+const request = require('request-promise');
+const cheerio = require('cheerio');
+
+request(prodLink)
+    .then(function (body) {
+        const $ = cheerio.load(body);
+        let price = $('#priceblock_ourprice').text().trim();
+        console.log(price);
+    })
+    .catch(function (err) {
+        console.log(err);
+    });
+
+```
+
+### Disk I/O
+
+Reading a file, especially a large one may take some time to complete. If we were to use a synchronous file reading function, the rest of your program wouldn't be able to run because it is stuck waiting for the file to be read. Instead, we can use asynchronous file reading functions which allow for background loading of the file, whilst keeping your program humming along.
+
+```javascript
+
+const fs = require('fs');
+
+function readFileWithPromise(filePath) {
+    return new Promise(function(resolve, reject) {
+        fs.readFile(filePath, 'utf8', function(err, data) {
+            if (err) {
+                reject(err);
+            } else {
+                resolve(data);
+            }
+        });
+    })
+
+};
+
+readFileWithPromise('path/to/file')
+    .then(data => console.log(data))
+    .catch(err => console.log(err));
+
+```
+
+
+## Benefits
```

Comment:
This section and the next section seem thin. One option is to get rid of them and absorb the content into the previous section. i.e., explain them as different 'ways of using promises' and mention the benefits at the same time. For example, the benefit of avoiding callback hell can be mentioned in the timeout example.